# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [248]:
# Import your libraries:

%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sqlalchemy import create_engine

import mysql.connector
import getpass
import pymysql
import sqlalchemy
sqlalchemy.__version__ 

from sklearn.model_selection import train_test_split


# Challenge 0 - Load,Query and Create connection of your Dataset

#### A lot of the times you won't have files already saved in an Excel or CSV file for you to prepare your data, implying that a lot of times you'll be extracting data from a Data Warehouse or Data Lake, majority of times through SQL.
#### A couple of times you may want to do some queries on your table in mySQL to get a slight view on what you have in hands,
#### so let's simulate that!

#### First we'll need to create a database and table in mySQL:

##### 1º- Open the austin_weather.sql file in MySQL Workbench and run the script into a desired schema.

#### 2º- As we are in mySQL Workbench, we can do some queries there, to get an overview on some characteristics of our data:
 - a) How many days are recorded in the dataset?
 - b) What is the day with the Highest Temperature in Fahrenheit (column TempHighF)
 - c) What is the average Humidity across all days? (column HumidityAvgPercent)
 - d) Top 10 days, where SeaLevelPressureAvgInches is the highest, knowing DewPointAvgF is higher than 28 ?

In [249]:
# Write your answers below:

# a) 1319
# b) 2017-07-29 00:00:00
# c) 66.57088703563305
# d) 
# 2016-11-19 00:00:00
# 2016-12-08 00:00:00
# 2014-12-31 00:00:00
# 2017-03-03 00:00:00
# 2016-01-01 00:00:00
# 2015-01-04 00:00:00
# 2014-01-23 00:00:00
# 2014-11-27 00:00:00
# 2015-12-04 00:00:00
# 2015-02-23 00:00:00

#### Now that you explored a couple of elements in your table, let's bring your table into this jupyter notebook, by creating a Python-SQL connection like you did on MySQL Project!
#### In case you need a little refresher check this [link](https://www.dataquest.io/blog/sql-insert-tutorial/).
##### 1º - Create a connection using sqlalchemy from python to mysql 
##### 2º- Load the table into a variable called weather_df


In [250]:
# Your code here
saved_pass = getpass.getpass()

In [251]:
connection = pymysql.connect(host= '127.0.0.1',
                         user='root',
                         password=saved_pass,
                         database='lab_week_17',
                         port=3306)

In [252]:
# Create cursor
my_cursor = connection.cursor()

In [253]:
# Execute Query
my_cursor.execute("SELECT * from austin_weather")

1319

In [254]:
# Fetch the records
result = my_cursor.fetchall()

In [255]:
column_names = [description[0] for description in my_cursor.description]
column_names

['Date',
 'TempHighF',
 'TempAvgF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'PrecipitationSumInches',
 'Events']

In [256]:
for i in result:
    
    print(i)

(datetime.datetime(2013, 12, 21, 0, 0), 74, 60, 45, '67', '49', '43', '93', '75', '57', '29.86', '29.68', '29.59', '10', '7', '2', '20', '4', '31', '0.46', 'Rain , Thunderstorm')
(datetime.datetime(2013, 12, 22, 0, 0), 56, 48, 39, '43', '36', '28', '93', '68', '43', '30.41', '30.13', '29.87', '10', '10', '5', '16', '6', '25', '0', ' ')
(datetime.datetime(2013, 12, 23, 0, 0), 58, 45, 32, '31', '27', '23', '76', '52', '27', '30.56', '30.49', '30.41', '10', '10', '10', '8', '3', '12', '0', ' ')
(datetime.datetime(2013, 12, 24, 0, 0), 61, 46, 31, '36', '28', '21', '89', '56', '22', '30.56', '30.45', '30.3', '10', '10', '7', '12', '4', '20', '0', ' ')
(datetime.datetime(2013, 12, 25, 0, 0), 58, 50, 41, '44', '40', '36', '86', '71', '56', '30.41', '30.33', '30.27', '10', '10', '7', '10', '2', '16', 'T', ' ')
(datetime.datetime(2013, 12, 26, 0, 0), 57, 48, 39, '39', '36', '33', '79', '63', '47', '30.47', '30.4', '30.34', '10', '9', '7', '12', '3', '17', '0', ' ')
(datetime.datetime(2013, 12, 

In [257]:
weather_df = pd.DataFrame(result, columns=column_names)

In [258]:
weather_df

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0    2013-12-21         74        60        45            67           49   
1    2013-12-22         56        48        39            43           36   
2    2013-12-23         58        45        32            31           27   
3    2013-12-24         61        46        31            36           28   
4    2013-12-25         58        50        41            44           40   
...         ...        ...       ...       ...           ...          ...   
1314 2017-07-27        103        89        75            71           67   
1315 2017-07-28        105        91        76            71           64   
1316 2017-07-29        107        92        77            72           64   
1317 2017-07-30        106        93        79            70           68   
1318 2017-07-31         99        88        77            66           61   

     DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0              43                  93                 75                 57   
1              28                  93                 68                 43   
2              23                  76                 52                 27   
3              21                  89                 56                 22   
4              36                  86                 71                 56   
...           ...                 ...                ...                ...   
1314           61                  82                 54                 25   
1315           55                  87                 54                 20   
1316           55                  82                 51                 19   
1317           63                  69                 48                 27   
1318           54                  64                 43                 22   

      ... SeaLevelPressureAvgInches SeaLevelPressureLowInches  \
0     ...                     29.68                     29.59   
1     ...                     30.13                     29.87   
2     ...                     30.49                     30.41   
3     ...                     30.45                      30.3   
4     ...                     30.33                     30.27   
...   ...                       ...                       ...   
1314  ...                     29.97                     29.88   
1315  ...                      29.9                     29.81   
1316  ...                     29.86                     29.79   
1317  ...                     29.91                     29.87   
1318  ...                     29.97                     29.91   

     VisibilityHighMiles VisibilityAvgMiles VisibilityLowMiles WindHighMPH  \
0                     10                  7                  2          20   
1                     10                 10                  5          16   
2                     10                 10                 10           8   
3                     10                 10                  7          12   
4                     10                 10                  7          10   
...                  ...                ...                ...         ...   
1314                  10                 10                 10          12   
1315                  10                 10                 10          14   
1316                  10                 10                 10          12   
1317                  10                 10                 10          13   
1318                  10                 10                 10          12   

     WindAvgMPH WindGustMPH PrecipitationSumInches               Events  
0             4          31                   0.46  Rain , Thunderstorm  
1             6          25                      0                       
2             3          12                      0                       
3             4          20                      0                       
4             2          16                      


# Challenge 1 - Describe the Dataset

#### Describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [259]:
# Your code here
weather_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Date                        1319 non-null   datetime64[ns]
 1   TempHighF                   1319 non-null   int64         
 2   TempAvgF                    1319 non-null   int64         
 3   TempLowF                    1319 non-null   int64         
 4   DewPointHighF               1319 non-null   object        
 5   DewPointAvgF                1319 non-null   object        
 6   DewPointLowF                1319 non-null   object        
 7   HumidityHighPercent         1319 non-null   object        
 8   HumidityAvgPercent          1319 non-null   object        
 9   HumidityLowPercent          1319 non-null   object        
 10  SeaLevelPressureHighInches  1319 non-null   object        
 11  SeaLevelPressureAvgInches   1319 non-null   object      

In [260]:
# Your code here
weather_df.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

In [261]:
# Your code here
weather_df[['TempHighF','TempAvgF','TempLowF']].describe()

TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [262]:
# Your code here
weather_df['DewPointAvgF'].unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [263]:
# Your observation here
# Type object due to we have '-' sign  in column. We need to find another way how to represent such data with missing values

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [264]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [265]:
# Your code here
for column in wrong_type_columns:
    weather_df[column] = pd.to_numeric(weather_df[column], errors ='coerce')


In [266]:
# weather_df[wrong_type_columns] = weather_df[wrong_type_columns].apply(pd.to_numeric, errors='coerce')

In [267]:
weather_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Date                        1319 non-null   datetime64[ns]
 1   TempHighF                   1319 non-null   int64         
 2   TempAvgF                    1319 non-null   int64         
 3   TempLowF                    1319 non-null   int64         
 4   DewPointHighF               1312 non-null   float64       
 5   DewPointAvgF                1312 non-null   float64       
 6   DewPointLowF                1312 non-null   float64       
 7   HumidityHighPercent         1317 non-null   float64       
 8   HumidityAvgPercent          1317 non-null   float64       
 9   HumidityLowPercent          1317 non-null   float64       
 10  SeaLevelPressureHighInches  1316 non-null   float64       
 11  SeaLevelPressureAvgInches   1316 non-null   float64     

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [268]:
# Your code here
missing_values = weather_df[weather_df.isnull().any(axis = 1)]
missing_values 

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
4    2013-12-25         58        50        41           44.0          40.0   
6    2013-12-27         60        53        45           41.0          39.0   
7    2013-12-28         62        51        40           43.0          39.0   
42   2014-02-01         76        66        55           62.0          59.0   
51   2014-02-10         60        48        35           49.0          36.0   
...         ...        ...       ...       ...            ...           ...   
1269 2017-06-12         94        85        75           72.0          68.0   
1288 2017-07-01         98        89        79           76.0          72.0   
1304 2017-07-17         98        88        77           75.0          71.0   
1308 2017-07-21        104        91        77           74.0          68.0   
1313 2017-07-26        103        91        78           75.0          70.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
4             36.0                 86.0                71.0   
6             37.0                 83.0                65.0   
7             33.0                 92.0                64.0   
42            41.0                 81.0                71.0   
51            30.0                 82.0                74.0   
...            ...                  ...                 ...   
1269          62.0                 87.0                62.0   
1288          66.0                 85.0                62.0   
1304          66.0                 88.0                63.0   
1308          61.0                 85.0                56.0   
1313          61.0                 91.0                58.0   

      HumidityLowPercent  ...  SeaLevelPressureAvgInches  \
4                   56.0  ...                      30.33   
6                   47.0  ...                      30.39   
7                   36.0  ...                      30.17   
42                  60.0  ...                      29.81   
51                  66.0  ...                      30.15   
...                  ...  ...                        ...   
1269                36.0  ...                      29.89   
1288                38.0  ...                      30.02   
1304                38.0  ...                      29.90   
1308                27.0  ...                      29.95   
1313                25.0  ...                      30.03   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
4                         30.27                 10.0                10.0   
6                         30.34                 10.0                 9.0   
7                         30.04                 10.0                10.0   
42                        29.75                 10.0                10.0   
51                        30.02                 10.0                 8.0   
...                         ...                  ...                 ...   
1269                      29.81                 10.0                10.0   
1288                      29.92                 10.0                 9.0   
1304                      29.84                 10.0                10.0   
1308                      29.86                 10.0                10.0   
1313                      29.93                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
4                    7.0         10.0         2.0         16.0   
6                    7.0          7.0         1.0         11.0   
7                    7.0         10.0         2.0         14.0   
42                   9.0         14.0         6.0         26.0   
51                   4.0         15.0         9.0         23.0   
...                  ...          ...         ...          ...   
1269                 2.0         15.0         7.0         22.0   
1288                 6.0         15.0         7.0         24.0   
1304                10.0         16.0         3.0         23.0   
1308       

In [269]:
missing_values.info()

<class 'pandas.core.frame.DataFrame'>
Index: 136 entries, 4 to 1313
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Date                        136 non-null    datetime64[ns]
 1   TempHighF                   136 non-null    int64         
 2   TempAvgF                    136 non-null    int64         
 3   TempLowF                    136 non-null    int64         
 4   DewPointHighF               129 non-null    float64       
 5   DewPointAvgF                129 non-null    float64       
 6   DewPointLowF                129 non-null    float64       
 7   HumidityHighPercent         134 non-null    float64       
 8   HumidityAvgPercent          134 non-null    float64       
 9   HumidityLowPercent          134 non-null    float64       
 10  SeaLevelPressureHighInches  133 non-null    float64       
 11  SeaLevelPressureAvgInches   133 non-null    float64       
 12

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [270]:
# Your code here
print(len(weather_df))
print(len(missing_values))

1319
136


#### Calculate the ratio of missing rows to total rows

In [271]:
# Your code here
ratio = len(missing_values)/len(weather_df)
ratio

0.10310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [272]:
# Your code here

columns = missing_values.columns.to_list()

for column in columns:
    na_values = missing_values[column].isna().sum()
    print (f'{column}:{na_values}')


Date:0
TempHighF:0
TempAvgF:0
TempLowF:0
DewPointHighF:7
DewPointAvgF:7
DewPointLowF:7
HumidityHighPercent:2
HumidityAvgPercent:2
HumidityLowPercent:2
SeaLevelPressureHighInches:3
SeaLevelPressureAvgInches:3
SeaLevelPressureLowInches:3
VisibilityHighMiles:12
VisibilityAvgMiles:12
VisibilityLowMiles:12
WindHighMPH:2
WindAvgMPH:2
WindGustMPH:4
PrecipitationSumInches:124
Events:0


#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [273]:
# Your code here
missing_values['PrecipitationSumInches'].isna().sum()/len(weather_df)

0.09401061410159212

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [274]:
# Your code here 

weather_df.drop(['PrecipitationSumInches'],axis=1, inplace = True)

# Print `austin` to confirm the column is indeed removed

weather_df

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0    2013-12-21         74        60        45           67.0          49.0   
1    2013-12-22         56        48        39           43.0          36.0   
2    2013-12-23         58        45        32           31.0          27.0   
3    2013-12-24         61        46        31           36.0          28.0   
4    2013-12-25         58        50        41           44.0          40.0   
...         ...        ...       ...       ...            ...           ...   
1314 2017-07-27        103        89        75           71.0          67.0   
1315 2017-07-28        105        91        76           71.0          64.0   
1316 2017-07-29        107        92        77           72.0          64.0   
1317 2017-07-30        106        93        79           70.0          68.0   
1318 2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH               E

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [275]:
# Your code here
austin_fixed = weather_df.interpolate(method='linear')

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [276]:
# Your code here
austin_fixed[austin_fixed.isnull().any(axis = 1)]

Empty DataFrame
Columns: [Date, TempHighF, TempAvgF, TempLowF, DewPointHighF, DewPointAvgF, DewPointLowF, HumidityHighPercent, HumidityAvgPercent, HumidityLowPercent, SeaLevelPressureHighInches, SeaLevelPressureAvgInches, SeaLevelPressureLowInches, VisibilityHighMiles, VisibilityAvgMiles, VisibilityLowMiles, WindHighMPH, WindAvgMPH, WindGustMPH, Events]
Index: []

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [277]:
# Your code here:
austin_fixed['Events'].value_counts()

Events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [278]:
# Your answer:
#We can see that biggest number of cells 903 relates to empty Event

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [279]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for event in event_list:
    austin_fixed[event]=0

# Print your new dataframe to check whether new columns have been created:

austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0    2013-12-21         74        60        45           67.0          49.0   
1    2013-12-22         56        48        39           43.0          36.0   
2    2013-12-23         58        45        32           31.0          27.0   
3    2013-12-24         61        46        31           36.0          28.0   
4    2013-12-25         58        50        41           44.0          40.0   
...         ...        ...       ...       ...            ...           ...   
1314 2017-07-27        103        89        75           71.0          67.0   
1315 2017-07-28        105        91        76           71.0          64.0   
1316 2017-07-29        107        92        77           72.0          64.0   
1317 2017-07-30        106        93        79           70.0          68.0   
1318 2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
0                   57.0  ...                 7.0                 2.0   
1                   43.0  ...                10.0                 5.0   
2                   27.0  ...                10.0                10.0   
3                   22.0  ...                10.0                 7.0   
4                   56.0  ...                10.0                 7.0   
...                  ...  ...                 ...                 ...   
1314                25.0  ...                10.0                10.0   
1315                20.0  ...                10.0                10.0   
1316                19.0  ...                10.0                10.0   
1317                27.0  ...                10.0                10.0   
1318                22.0  ...                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH               Events  Snow  Fog  \
0            20.0         4.0         31.0  Rain , Thunderstorm     0    0   
1            16.0         6.0         25.0                          0    0   
2             8.0         3.0         12.0                          0    0   
3            12.0         4.0         20.0                          0    0   
4            10.0         2.0         16.0                          0    0   
...           ...         ...          ...                  ...   ...  ...   
1314         12.0         5.0         21.0                          0    0   
1315         14.0         5.0         20.0                          0    0   
1316         12.0         4.0         17.0                          0    0   
1317         13.0         4.0         20.0                          0    0   
1318         12.0         4.0         20.0                          0    0   

      Rain Thunderstorm  
0        0            0  
1        0            0  
2        0            0  
3        0            0  
4        0            0  
...    ...          ...  
1314     0            0  
1315     0            0  
1316     0            0  
1317     0            0  
1318     0            0  

[1319 rows x 24 columns]

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [280]:
# Your code here
austin_fixed['Rain'] = austin_fixed['Events'].str.contains('Rain').astype(int)
austin_fixed['Snow'] = austin_fixed['Events'].str.contains('Snow').astype(int)
austin_fixed['Fog'] = austin_fixed['Events'].str.contains('Fog').astype(int)
austin_fixed['Thunderstorm'] = austin_fixed['Events'].str.contains('Thunderstorm').astype(int)

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [281]:
# Your code here
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0    2013-12-21         74        60        45           67.0          49.0   
1    2013-12-22         56        48        39           43.0          36.0   
2    2013-12-23         58        45        32           31.0          27.0   
3    2013-12-24         61        46        31           36.0          28.0   
4    2013-12-25         58        50        41           44.0          40.0   
...         ...        ...       ...       ...            ...           ...   
1314 2017-07-27        103        89        75           71.0          67.0   
1315 2017-07-28        105        91        76           71.0          64.0   
1316 2017-07-29        107        92        77           72.0          64.0   
1317 2017-07-30        106        93        79           70.0          68.0   
1318 2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
0                   57.0  ...                 7.0                 2.0   
1                   43.0  ...                10.0                 5.0   
2                   27.0  ...                10.0                10.0   
3                   22.0  ...                10.0                 7.0   
4                   56.0  ...                10.0                 7.0   
...                  ...  ...                 ...                 ...   
1314                25.0  ...                10.0                10.0   
1315                20.0  ...                10.0                10.0   
1316                19.0  ...                10.0                10.0   
1317                27.0  ...                10.0                10.0   
1318                22.0  ...                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH               Events  Snow  Fog  \
0            20.0         4.0         31.0  Rain , Thunderstorm     0    0   
1            16.0         6.0         25.0                          0    0   
2             8.0         3.0         12.0                          0    0   
3            12.0         4.0         20.0                          0    0   
4            10.0         2.0         16.0                          0    0   
...           ...         ...          ...                  ...   ...  ...   
1314         12.0         5.0         21.0                          0    0   
1315         14.0         5.0         20.0                          0    0   
1316         12.0         4.0         17.0                          0    0   
1317         13.0         4.0         20.0                          0    0   
1318         12.0         4.0         20.0                          0    0   

      Rain Thunderstorm  
0        1            1  
1        0            0  
2        0            0  
3        0            0  
4        0            0  
...    ...          ...  
1314     0            0  
1315     0            0  
1316     0            0  
1317     0            0  
1318     0            0  

[1319 rows x 24 columns]

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [282]:
# Your code here
austin_fixed.drop(['Events'],axis = 1, inplace = True)

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [283]:
# Your code here

from datetime import datetime

In [284]:
def convert_date_to_ordinal(date):
    return date.toordinal()

# Apply the function to the 'Date' column
austin_fixed['Date'] = austin_fixed['Date'].apply(convert_date_to_ordinal)

#### Print `austin_fixed` to check your `Date` column.

In [285]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0  ...                 10.0                 7.0                 2.0   
1  ...                 10.0                10.0                 5.0   
2  ...                 10.0                10.0                10.0   
3  ...                 10.0                10.0                 7.0   
4  ...                 10.0                10.0                 7.0   

   WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0         20.0         4.0         31.0     0    0     1             1  
1         16.0         6.0         25.0     0    0     0             0  
2          8.0         3.0         12.0     0    0     0             0  
3         12.0         4.0         20.0     0    0     0             0  
4         10.0         2.0         16.0     0    0     0             0  

[5 rows x 23 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [286]:
# Your code here:
X =austin_fixed.drop(columns = ['TempAvgF'])
y = austin_fixed['TempAvgF']
X_train, X_test,y_train, y_test = train_test_split(X,y)

print(X_train.shape)
print(X_test.shape)
X_train.head()

(989, 22)
(330, 22)


Date  TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
108  735331         77        50           50.0          38.0          27.0   
39   735262         49        27           19.0          13.0           9.0   
367  735590         59        43           44.0          40.0          34.0   
123  735346         87        59           64.0          61.0          57.0   
609  735832         99        75           75.0          73.0          65.0   

     HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
108                 96.0                57.0                18.0   
39                  63.0                42.0                21.0   
367                 77.0                66.0                55.0   
123                 90.0                68.0                46.0   
609                 94.0                65.0                35.0   

     SeaLevelPressureHighInches  ...  VisibilityHighMiles  VisibilityAvgMiles  \
108                       30.20  ...                 10.0                10.0   
39                        30.52  ...                 10.0                10.0   
367                       30.02  ...                 10.0                10.0   
123                       29.93  ...                 10.0                10.0   
609                       29.98  ...                 10.0                10.0   

     VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
108                10.0         22.0         8.0         37.0     0    0   
39                 10.0         15.0         5.0         23.0     0    0   
367                10.0         18.0         8.0         31.0     0    0   
123                10.0         17.0         7.0         29.0     0    0   
609                10.0         13.0         6.0         22.0     0    0   

     Rain  Thunderstorm  
108     0             0  
39      0             0  
367     1             0  
123     0             0  
609     0             0  

[5 rows x 22 columns]

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [287]:
#Your code here:


Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [288]:
#Your code here:


#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [292]:
# Your code here:
import math

# Total number of rows in the dataset
total_rows = len(austin_fixed)

# Calculate 80% of the total rows and round it up to the next integer
ts_rows = math.ceil(0.8 * total_rows)

# Display the result
print(total_rows, ts_rows)

1319 1056


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [295]:
# Your code here:

X_ts_train = X.iloc[:ts_rows]   # First `ts_rows` rows
X_ts_test = X.iloc[ts_rows:]

print(len(X_ts_train), len(X_ts_test))


1056 263


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [296]:
# Your code here:
y_ts_train = y.iloc[:ts_rows]   # First `ts_rows` rows
y_ts_test = y.iloc[ts_rows:]

print(len(y_ts_train), len(y_ts_test))


1056 263
